src/port1.0/portconfigure.tcl

```diff
@@ -528,98 +599,471 @@ proc portconfigure::configure_get_default_compiler {} {
+            lappend compilers $cc
+        }
+    }
+    return $compilers
+}
+# utility procedure: get Clang compilers based on os.major
+proc portconfigure::get_clang_compilers {} {
+    global os.major porturl
+    set compilers ""
+    set compiler_file [getportresourcepath $porturl "port1.0/compilers/clang_compilers.tcl"]
+    if {[file exists ${compiler_file}]} {
+        source ${compiler_file}
     } else {
-        set our_stdlib $cxx_stdlib
+        ui_error "Clang compilers file not found"
+        return -code error "${compiler_file} does not exist"
     }
-    if {$our_stdlib eq "libc++"} {
-        # clang-3.5+ require libc++
-        lappend compilers macports-clang-5.0 macports-clang-4.0
-
-        if {${os.major} < 17} {
-            # The High Sierra SDK requires a toolchain that can apply nullability to uuid_t
-            lappend compilers macports-clang-3.9
+    return ${compilers}
+}
+# utility procedure: get GCC compilers based on os.major
+proc portconfigure::get_gcc_compilers {} {
+    global os.major porturl
+    set compilers ""
+    set compiler_file [getportresourcepath $porturl "port1.0/compilers/gcc_compilers.tcl"]
+    if {[file exists ${compiler_file}]} {
+        source ${compiler_file}
+    } else {
+        ui_error "GCC compilers file not found"
+        return -code error "${compiler_file} does not exist"
+    }
+    return ${compilers}
+}
+# utility procedure: get MPI wrapper for a given compiler
+proc portconfigure::get_mpi_wrapper {mpi compiler} {
+    set parts [split ${compiler} -]
+    if {[lindex ${parts} 0] ne "macports"} {
+        return macports-${mpi}-default
+    } else {
+        set type [lindex ${parts} 1]
+        set ver  [lindex ${parts} 2]
+        if {${type} eq "clang" && [vercmp ${ver} 3.3] < 0} {
+            return ""
         }
-
-        if {${os.major} < 16} {
-            # The Sierra SDK requires a toolchain that supports class properties
-            lappend compilers macports-clang-3.7
+        if {${type} eq "gcc" && [vercmp ${ver} 4.3] < 0} {
+            return ""
+        }
+        if {${type} eq "g95"} {
+            return ""
         }
+        return macports-${mpi}-${type}-${ver}
     }
+}
+# utility procedure: get system compiler version by running it
+proc compiler.command_line_tools_version {compiler} {
+    switch ${compiler} {
+        clang {
+            set re {clang(?:_.*)?-([0-9.]+)}
+        }
+        llvm-gcc-4.2 {
+            set re {LLVM build ([0-9.]+)}
+        }
+        gcc-4.2 {
+            set re {build ([0-9.]+)}
+        }
+        gcc-4.0 -
+        apple-gcc-4.2 {
+            set re {build ([0-9.]+)}
+        }
+        default {
+            return -code error "don't know how to determine build number of compiler \"${compiler}\""
+        }
+    }
+    set cc [portconfigure::configure_get_compiler cc ${compiler}]
+    if {![file executable ${cc}]} return
+    if {[catch {regexp ${re} [exec ${cc} -v 2>@1] -> compiler_version}]} return
```

Comment:
Could this value be cached? Preferably outside the portfile context so it doesn't have to be re-evaluated for each port?

Reply:
I am afraid I do not know how to do that.
This code came from the [compiler blacklist PG](https://github.com/macports/macports-ports/blob/master/_resources/port1.0/group/compiler_blacklist_versions-1.0.tcl).